File: dotnet/ArcticDB/ArcticLibrary.cs
// Copyright 2026 Man Group Operations Limited
//
// Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source License, use of this software
// will be governed by the Apache License, version 2.0.

using System.Runtime.InteropServices;

namespace ArcticDB;

/// <summary>
/// High-level wrapper around the ArcticDB C API.
/// Implements <see cref="IDisposable"/> for deterministic resource cleanup.
/// </summary>
/// <example>
/// <code>
/// using var lib = ArcticLibrary.OpenLmdb("/tmp/test_db");
/// lib.WriteTestData("prices", 1000, 5);
/// var result = lib.ReadStream("prices");
/// Console.WriteLine($"Read {result.TotalRows} rows");
/// </code>
/// </example>
public class ArcticLibrary : IDisposable
{
    private IntPtr _handle;
    private bool _disposed;

    private ArcticLibrary(IntPtr handle)
    {
        _handle = handle;
    }

    /// <summary>
    /// Open an LMDB-backed ArcticDB library at the given path.
    /// </summary>
    /// <param name="path">Filesystem path for LMDB storage (created if absent).</param>
    /// <returns>A new ArcticLibrary instance (caller must dispose).</returns>
    public static ArcticLibrary OpenLmdb(string path)
    {
        var err = new ArcticNative.ArcticError();
        int rc = ArcticNative.arctic_library_open_lmdb(path, out IntPtr handle, ref err);
        ArcticNative.CheckError(rc, ref err);
        return new ArcticLibrary(handle);
    }

    /// <summary>
    /// Write synthetic test data: a timeseries-indexed DataFrame with float64 columns.
    /// </summary>
    /// <param name="symbol">Symbol name.</param>
    /// <param name="numRows">Number of rows.</param>
    /// <param name="numColumns">Number of float64 columns (named col_0..col_N).</param>
    public void WriteTestData(string symbol, long numRows, long numColumns)
    {
        var err = new ArcticNative.ArcticError();
        int rc = ArcticNative.arctic_write_test_data(_handle, symbol, numRows, numColumns, ref err);
        ArcticNative.CheckError(rc, ref err);
    }

    /// <summary>
    /// Read the latest version of a symbol as a streaming Arrow result.
    /// </summary>
    public ReadResult ReadStream(string symbol) => ReadStream(symbol, -1);

    /// <summary>
    /// Read a specific version of a symbol as a streaming Arrow result.
    /// </summary>
    /// <param name="symbol">Symbol name.</param>
    /// <param name="version">Version number, or -1 for latest.</param>
    /// <returns>Summary of the data read.</returns>
    public ReadResult ReadStream(string symbol, long version)
    {
        var stream = new ArcticNative.ArcticArrowArrayStream();
        var err = new ArcticNative.ArcticError();
        int rc = ArcticNative.arctic_read_stream(_handle, symbol, version, ref stream, ref err);
        ArcticNative.CheckError(rc, ref err);

        try
        {
            // 1. Get schema
            var schema = new ArcticNative.ArrowSchema();
            var getSchema = Marshal.GetDelegateForFunctionPointer<ArcticNative.GetSchemaDelegate>(stream.GetSchema);
            int schemaRc = getSchema(ref stream, ref schema);
            if (schemaRc != 0)
                throw new ArcticException(schemaRc, "get_schema failed");

            // Read column names from schema children
            var columnNames = new List<string>();
            if (schema.NChildren > 0 && schema.Children != IntPtr.Zero)
            {
                for (long i = 0; i < schema.NChildren; i++)
                {
                    IntPtr childPtr = Marshal.ReadIntPtr(schema.Children, (int)(i * IntPtr.Size));
                    if (childPtr != IntPtr.Zero)
                    {
                        var child = Marshal.PtrToStructure<ArcticNative.ArrowSchema>(childPtr);
                        if (child.Name != IntPtr.Zero)
                        {
                            string? name = Marshal.PtrToStringUTF8(child.Name);
                            if (name != null)
                                columnNames.Add(name);
                        }
                    }
                }
            }

            // Release schema
            if (schema.Release != IntPtr.Zero)
            {
                var releaseSchema = Marshal.GetDelegateForFunctionPointer<ArcticNative.ReleaseArrowDelegate>(schema.Release);
                releaseSchema(ref schema);
            }

            // 2. Consume batches
            long totalRows = 0;
            int batchCount = 0;

            var getNext = Marshal.GetDelegateForFunctionPointer<ArcticNative.GetNextDelegate>(stream.GetNext);

            while (true)
            {
                var array = new ArcticNative.ArrowArray();
                int nextRc = getNext(ref stream, ref array);
                if (nextRc != 0)
                    throw new ArcticException(nextRc, "get_next failed");

                // release == NULL means end of stream
                if (array.Release == IntPtr.Zero)
                    break;

                totalRows += array.Length;
                batchCount++;

                // Release this array
                var releaseArray = Marshal.GetDelegateForFunctionPointer<ArcticNative.ReleaseArrayDelegate>(array.Release);
                releaseArray(ref array);
            }

            return new ReadResult(columnNames, totalRows, batchCount);
        }
        finally
        {
            // 3. Release stream
            if (stream.Release != IntPtr.Zero)
            {
                var releaseStream = Marshal.GetDelegateForFunctionPointer<ArcticNative.ReleaseStreamDelegate>(stream.Release);
                releaseStream(ref stream);
            }
        }
    }

    /// <summary>
    /// List all symbols in this library.
    /// </summary>
    public List<string> ListSymbols()
    {
        var err = new ArcticNative.ArcticError();
        int rc = ArcticNative.arctic_list_symbols(_handle, out IntPtr symbolsPtr, out long count, ref err);
        ArcticNative.CheckError(rc, ref err);

        var result = new List<string>();
        if (count > 0 && symbolsPtr != IntPtr.Zero)
        {
            for (long i = 0; i < count; i++)
            {
                IntPtr strPtr = Marshal.ReadIntPtr(symbolsPtr, (int)(i * IntPtr.Size));
                string? s = Marshal.PtrToStringUTF8(strPtr);
                if (s != null) result.Add(s);
            }
            ArcticNative.arctic_free_symbols(symbolsPtr, count);
        }

        return result;
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            if (_handle != IntPtr.Zero)
            {
                ArcticNative.arctic_library_close(_handle);
                _handle = IntPtr.Zero;
            }
            GC.SuppressFinalize(this);
        }
    }

    ~ArcticLibrary()
    {
        Dispose();
    }
}

/// <summary>
/// Summary of data read from an Arrow stream.
/// </summary>
/// <param name="ColumnNames">Names of data columns (excludes the index).</param>
/// <param name="TotalRows">Total number of rows across all batches.</param>
/// <param name="BatchCount">Number of Arrow record batches consumed.</param>
public record ReadResult(
    List<string> ColumnNames,
    long TotalRows,
    int BatchCount
);

File: java/src/main/java/com/arcticdb/ArcticLibrary.java
/* Copyright 2026 Man Group Operations Limited
 *
 * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
 *
 * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software
 * will be governed by the Apache License, version 2.0.
 */

package com.arcticdb;

import java.lang.foreign.*;
import java.util.ArrayList;
import java.util.List;

/**
 * High-level wrapper around the ArcticDB C API.
 *
 * <p>Implements {@link AutoCloseable} for deterministic resource cleanup.
 * Each instance owns a confined {@link Arena} that is closed when the library is closed.
 *
 * <pre>{@code
 * try (var lib = ArcticLibrary.openLmdb("/tmp/test_db")) {
 *     lib.writeTestData("prices", 1000, 5);
 *     ReadResult result = lib.readStream("prices");
 *     System.out.println("Read " + result.totalRows() + " rows");
 * }
 * }</pre>
 */
public class ArcticLibrary implements AutoCloseable {

    private final Arena arena;
    private final MemorySegment handle;

    private ArcticLibrary(Arena arena, MemorySegment handle) {
        this.arena = arena;
        this.handle = handle;
    }

    /**
     * Open an LMDB-backed ArcticDB library at the given path.
     *
     * @param path filesystem path for LMDB storage (created if absent)
     * @return a new ArcticLibrary instance (caller must close)
     */
    public static ArcticLibrary openLmdb(String path) {
        Arena arena = Arena.ofConfined();
        try {
            MemorySegment pathSeg = arena.allocateUtf8String(path);
            MemorySegment outPtr = arena.allocate(ValueLayout.ADDRESS);
            MemorySegment errSeg = arena.allocate(ArcticNative.ARCTIC_ERROR_LAYOUT);

            int rc = (int) ArcticNative.LIBRARY_OPEN_LMDB.invokeExact(pathSeg, outPtr, errSeg);
            ArcticNative.checkError(rc, errSeg);

            MemorySegment handle = outPtr.get(ValueLayout.ADDRESS, 0);
            return new ArcticLibrary(arena, handle);
        } catch (RuntimeException e) {
            arena.close();
            throw e;
        } catch (Throwable t) {
            arena.close();
            throw new RuntimeException(t);
        }
    }

    /**
     * Write synthetic test data: a timeseries-indexed DataFrame with float64 columns.
     *
     * @param symbol     symbol name
     * @param numRows    number of rows
     * @param numColumns number of float64 columns (named col_0..col_N)
     */
    public void writeTestData(String symbol, long numRows, long numColumns) {
        try (Arena local = Arena.ofConfined()) {
            MemorySegment symSeg = local.allocateUtf8String(symbol);
            MemorySegment errSeg = local.allocate(ArcticNative.ARCTIC_ERROR_LAYOUT);

            int rc = (int) ArcticNative.WRITE_TEST_DATA.invokeExact(handle, symSeg, numRows, numColumns, errSeg);
            ArcticNative.checkError(rc, errSeg);
        } catch (RuntimeException e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Read the latest version of a symbol as a streaming Arrow result.
     */
    public ReadResult readStream(String symbol) {
        return readStream(symbol, -1);
    }

    /**
     * Read a specific version of a symbol as a streaming Arrow result.
     *
     * @param symbol  symbol name
     * @param version version number, or -1 for latest
     * @return summary of the data read (column names, row count, batch count)
     */
    public ReadResult readStream(String symbol, long version) {
        try (Arena local = Arena.ofConfined()) {
            MemorySegment symSeg = local.allocateUtf8String(symbol);
            MemorySegment streamSeg = local.allocate(ArcticNative.STREAM_LAYOUT);
            MemorySegment errSeg = local.allocate(ArcticNative.ARCTIC_ERROR_LAYOUT);

            int rc = (int) ArcticNative.READ_STREAM.invokeExact(handle, symSeg, version, streamSeg, errSeg);
            ArcticNative.checkError(rc, errSeg);

            try {
                // 1. Get schema
                MemorySegment schemaSeg = local.allocate(ArcticNative.ARROW_SCHEMA_LAYOUT);
                int schemaRc = ArcticNative.callGetSchema(streamSeg, schemaSeg);
                if (schemaRc != 0) {
                    throw new RuntimeException("get_schema failed with code " + schemaRc);
                }

                // Read column names from schema children
                long nChildren = schemaSeg.get(ValueLayout.JAVA_LONG, 32); // n_children offset
                MemorySegment childrenPtr = schemaSeg.get(ValueLayout.ADDRESS, 40); // children offset
                List<String> columnNames = new ArrayList<>();

                if (nChildren > 0 && !childrenPtr.equals(MemorySegment.NULL)) {
                    // children is ArrowSchema**, an array of pointers
                    MemorySegment childrenArray = childrenPtr.reinterpret(nChildren * ValueLayout.ADDRESS.byteSize());
                    for (long i = 0; i < nChildren; i++) {
                        MemorySegment childPtr = childrenArray.get(ValueLayout.ADDRESS, i * ValueLayout.ADDRESS.byteSize());
                        if (!childPtr.equals(MemorySegment.NULL)) {
                            MemorySegment child = childPtr.reinterpret(ArcticNative.ARROW_SCHEMA_LAYOUT.byteSize());
                            MemorySegment namePtr = child.get(ValueLayout.ADDRESS, 8); // name offset
                            if (!namePtr.equals(MemorySegment.NULL)) {
                                columnNames.add(namePtr.reinterpret(256).getUtf8String(0));
                            }
                        }
                    }
                }

                // Release schema
                ArcticNative.callArrowRelease(schemaSeg, 56); // release at offset 56

                // 2. Consume batches
                long totalRows = 0;
                int batchCount = 0;

                while (true) {
                    MemorySegment arraySeg = local.allocate(ArcticNative.ARROW_ARRAY_LAYOUT);
                    int nextRc = ArcticNative.callGetNext(streamSeg, arraySeg);
                    if (nextRc != 0) {
                        throw new RuntimeException("get_next failed with code " + nextRc);
                    }

                    // Check if release is NULL → end of stream
                    MemorySegment releasePtr = arraySeg.get(ValueLayout.ADDRESS, 64); // release at offset 64
                    if (releasePtr.equals(MemorySegment.NULL)) {
                        break;
                    }

                    long length = arraySeg.get(ValueLayout.JAVA_LONG, 0); // length at offset 0
                    totalRows += length;
                    batchCount++;

                    // Release this array
                    ArcticNative.callArrowRelease(arraySeg, 64);
                }

                return new ReadResult(columnNames, totalRows, batchCount);
            } finally {
                // 3. Release stream
                ArcticNative.callStreamRelease(streamSeg);
            }
        } catch (RuntimeException e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * List all symbols in this library.
     */
    public List<String> listSymbols() {
        try (Arena local = Arena.ofConfined()) {
            MemorySegment outSymbols = local.allocate(ValueLayout.ADDRESS);
            MemorySegment outCount = local.allocate(ValueLayout.JAVA_LONG);
            MemorySegment errSeg = local.allocate(ArcticNative.ARCTIC_ERROR_LAYOUT);

            int rc = (int) ArcticNative.LIST_SYMBOLS.invokeExact(handle, outSymbols, outCount, errSeg);
            ArcticNative.checkError(rc, errSeg);

            long count = outCount.get(ValueLayout.JAVA_LONG, 0);
            MemorySegment symbolsArray = outSymbols.get(ValueLayout.ADDRESS, 0);

            List<String> result = new ArrayList<>();
            if (count > 0 && !symbolsArray.equals(MemorySegment.NULL)) {
                MemorySegment arr = symbolsArray.reinterpret(count * ValueLayout.ADDRESS.byteSize());
                for (long i = 0; i < count; i++) {
                    MemorySegment strPtr = arr.get(ValueLayout.ADDRESS, i * ValueLayout.ADDRESS.byteSize());
                    result.add(strPtr.reinterpret(256).getUtf8String(0));
                }
                // Free the native symbol list
                ArcticNative.FREE_SYMBOLS.invokeExact(symbolsArray, count);
            }

            return result;
        } catch (RuntimeException e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    @Override
    public void close() {
        try {
            ArcticNative.LIBRARY_CLOSE.invokeExact(handle);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        } finally {
            arena.close();
        }
    }

    /**
     * Summary of data read from an Arrow stream.
     *
     * @param columnNames names of data columns (excludes the index)
     * @param totalRows   total number of rows across all batches
     * @param batchCount  number of Arrow record batches consumed
     */
    public record ReadResult(
            List<String> columnNames,
            long totalRows,
            int batchCount
    ) {}
}
